scripts/street_sign_recognizer.py

```diff
@@ -18,47 +19,127 @@ def __init__(self):
+        """ A callback function to handle the OpenCV slider to select the saturation upper bound """
+        self.sat_upper_bound = val
+    def set_dilate_erode_iterations(self, val):
+        """ A callback function to handle the OpenCV slider to select the number of times to dilate and erode"""
+        self.dilate_erode_iterations = val
 
     def process_image(self, msg):
         """ Process image messages from ROS and stash them in an attribute
-            called cv_image for subsequent processing """
+            called cv_image. Converts to binary, erodes, and dilates for cropping. Calls cropping, then passes to matcher"""
+
         self.cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
+        kernel = np.ones((5,5),np.uint8)
+
+        #filter image
+        self.hsv_image = cv2.cvtColor(self.cv_image, cv2.COLOR_BGR2HSV) #change from rgb to hsv
+        self.binary_image = cv2.inRange(self.hsv_image, (self.hue_lower_bound,self.sat_lower_bound,self.value_lower_bound), (self.hue_upper_bound,self.sat_upper_bound,self.value_upper_bound))
+        self.binary_image = cv2.erode(self.binary_image, kernel, iterations=self.dilate_erode_iterations)
+        self.binary_image = cv2.dilate(self.binary_image, kernel, iterations=self.dilate_erode_iterations)
 
         left_top, right_bottom = self.sign_bounding_box()
         left, top = left_top
         right, bottom = right_bottom
 
         # crop bounding box region of interest
-        cropped_sign = self.cv_image[top:bottom, left:right]
+        cropped_sign = cv2.cvtColor(self.cv_image[top:bottom, left:right], cv2.COLOR_BGR2GRAY)
 
         # draw bounding box rectangle
         cv2.rectangle(self.cv_image, left_top, right_bottom, color=(0, 0, 255), thickness=5)
 
+        self.pred = self.matcher.predict(cropped_sign) #predict using template matcher
+
+
+
     def sign_bounding_box(self):
         """
-        Returns
-        -------
-        (left_top, right_bottom) where left_top and right_bottom are tuples of (x_pixel, y_pixel)
-            defining topleft and bottomright corners of the bounding box
+        Sections up image into grid, analyzes grid sections, uses grid to return (left_top, right_bottom) where left_top and right_bottom are tuples of (x_pixel, y_pixel)
+        defining topleft and bottomright corners of the bounding box
         """
-        # TODO: YOUR SOLUTION HERE
-        left_top = (200, 200)
-        right_bottom = (400, 400)
+        #how big sections of grid are
+        section_height = 30
+        section_width = 40
+        #lists to contain section with stuff in them
+        x_range = []
+        y_range = []
+        #slice into grid
+        for height in range(0,480,section_height):
+            for width in range(0,640,section_width):
+                img_slice = self.binary_image[height:height+section_height,width:width+section_width]
+                white_count = cv2.countNonZero(img_slice)
```

Comment:
Didn't know about this function!  Will use next time.

Another way to do it could be to do img_slice.sum().  Since your binary image is made up of True and False values, the sum of all the True values will be the same as countNonZero.  Just the "numpy" way of doing it for reference.